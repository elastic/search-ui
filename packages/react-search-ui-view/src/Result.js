import PropTypes from "prop-types";
import React from "react";

function Result({ fields, onClickLink, title, url }) {
  return (
    <li className="sui-result">
      <div className="sui-result__header">
        {title && !url && (
          <span
            className="sui-result__title"
            dangerouslySetInnerHTML={{ __html: title }}
          />
        )}
        {title && url && (
          <a
            className="sui-result__title sui-result__title-link"
            dangerouslySetInnerHTML={{ __html: title }}
            href={url}
            onClick={onClickLink}
            target="_blank"
            rel="noopener noreferrer"
          />
        )}
      </div>
      <div className="sui-result__body">
        <ul className="sui-result__details">
          {Object.keys(fields).map(key => (
            <li key={key}>
              <span className="sui-result__key">{key}</span>{" "}
              <span
                className="sui-result__value"
                dangerouslySetInnerHTML={{ __html: fields[key] }}
              />
            </li>
          ))}
        </ul>
      </div>
    </li>
  );
}

Result.propTypes = {
  fields: PropTypes.object.isRequired,
  onClickLink: PropTypes.func.isRequired,
  title: PropTypes.string,
  url: PropTypes.string
};

export default Result;
