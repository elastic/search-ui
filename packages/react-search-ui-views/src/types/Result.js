import PropTypes from "prop-types";

import FieldValueWrapper from "./FieldValueWrapper";

// Typically an object where keys are the field names, and values are field values.
// Also could be literally any other arbitrary value depending on the particular Search API response.
// Default views in Search UI know what to do with FieldValueWrapper values, but not arbitrary values, so it
// is usually better to work with FieldValueWrapper values. We encourage FieldValueWrapper, but we accept
// anything because we don't want users to have type error warnings unnecessarily.
//
// An example would be if a user requests "grouping" in an App Search API request. That will come back
// as "_group: {..}". It *should* be there in the Result so that a developer has it available to work
// with.
export default PropTypes.objectOf(
  PropTypes.oneOfType([PropTypes.any, FieldValueWrapper])
);
