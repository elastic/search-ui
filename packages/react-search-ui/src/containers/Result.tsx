import React from "react";
import { Component } from "react";
import {
  Result,
  ResultContainerProps,
  ResultContainerContext,
  ResultViewProps
} from "@elastic/react-search-ui-views";

import { withSearch } from "..";

export class ResultContainer extends Component<ResultContainerProps> {
  static defaultProps = {
    clickThroughTags: [],
    shouldTrackClickThrough: true
  };

  handleClickLink = (id) => {
    const { clickThroughTags, shouldTrackClickThrough, trackClickThrough } =
      this.props;

    if (shouldTrackClickThrough) {
      trackClickThrough(id, clickThroughTags);
    }
  };

  render() {
    const {
      className,
      result,
      titleField,
      urlField,
      thumbnailField,
      view,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      trackClickThrough,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      shouldTrackClickThrough,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      clickThroughTags,
      ...rest
    } = this.props;
    const View = view || Result;
    const id = result.id.raw;

    const viewProps: ResultViewProps = {
      className,
      result: result,
      key: `result-${id}`,
      onClickLink: () => this.handleClickLink(id),
      titleField,
      urlField,
      thumbnailField,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<ResultContainerProps, ResultContainerContext>(
  ({ trackClickThrough }) => ({ trackClickThrough })
)(ResultContainer);
