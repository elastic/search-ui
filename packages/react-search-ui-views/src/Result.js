import PropTypes from "prop-types";
import React from "react";

import { appendClassName, getUrlSanitizer } from "./view-helpers";
import { isFieldValueWrapper } from "./types/FieldValueWrapper";

function getFieldType(result, field, type) {
  if (result[field]) return result[field][type];
}

function getRaw(result, field) {
  return getFieldType(result, field, "raw");
}

function getSnippet(result, field) {
  return getFieldType(result, field, "snippet");
}

function htmlEscape(str) {
  if (!str) return "";

  return String(str)
    .replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

function getEscapedField(result, field) {
  // Fallback to raw values here, because non-string fields
  // will not have a snippet fallback. Raw values MUST be html escaped.
  const safeField =
    getSnippet(result, field) || htmlEscape(getRaw(result, field));
  return Array.isArray(safeField) ? safeField.join(", ") : safeField;
}

function getEscapedFields(result) {
  return Object.keys(result).reduce((acc, field) => {
    // If we receive an arbitrary value from the response, we may not properly
    // handle it, so we should filter out arbitrary values here.
    //
    // I.e.,
    // Arbitrary value: "_metaField: '1939191'"
    // vs.
    // FieldValueWrapper: "_metaField: {raw: '1939191'}"
    if (!isFieldValueWrapper(result[field])) return acc;
    return { ...acc, [field]: getEscapedField(result, field) };
  }, {});
}

function Result({
  className,
  result,
  onClickLink,
  titleField,
  urlField,
  ...rest
}) {
  const fields = getEscapedFields(result);
  const title = getEscapedField(result, titleField);
  const url = getUrlSanitizer(URL, location)(getRaw(result, urlField));

  return (
    <li className={appendClassName("sui-result", className)} {...rest}>
      <div className="sui-result__header">
        {title && !url && (
          <span
            className="sui-result__title"
            dangerouslySetInnerHTML={{ __html: title }}
          />
        )}
        {title && url && (
          <a
            className="sui-result__title sui-result__title-link"
            dangerouslySetInnerHTML={{ __html: title }}
            href={url}
            onClick={onClickLink}
            target="_blank"
            rel="noopener noreferrer"
          />
        )}
      </div>
      <div className="sui-result__body">
        <ul className="sui-result__details">
          {Object.entries(fields).map(([fieldName, fieldValue]) => (
            <li key={fieldName}>
              <span className="sui-result__key">{fieldName}</span>{" "}
              <span
                className="sui-result__value"
                dangerouslySetInnerHTML={{ __html: fieldValue }}
              />
            </li>
          ))}
        </ul>
      </div>
    </li>
  );
}

Result.propTypes = {
  result: PropTypes.object.isRequired,
  onClickLink: PropTypes.func.isRequired,
  className: PropTypes.string,
  titleField: PropTypes.string,
  urlField: PropTypes.string
};

export default Result;
