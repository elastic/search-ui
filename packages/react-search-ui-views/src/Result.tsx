import React from "react";

import {
  appendClassName,
  getUrlSanitizer,
  formatResult,
  getEscapedField,
  getRaw
} from "./view-helpers";
import type { SearchContextState, SearchResult } from "@elastic/search-ui";
import { BaseContainerProps } from "./types";

export type ResultContainerContext = Pick<
  SearchContextState,
  "trackClickThrough"
>;

export type ResultContainerProps = BaseContainerProps &
  ResultContainerContext & {
    view?: React.ComponentType<ResultViewProps>;
    clickThroughTags?: string[];
    titleField?: string;
    urlField?: string;
    thumbnailField?: string;
    result: SearchResult;
    shouldTrackClickThrough?: boolean;
  };

export type ResultViewProps = BaseContainerProps &
  Pick<
    ResultContainerProps,
    "result" | "titleField" | "urlField" | "thumbnailField"
  > & {
    key?: string;
    onClickLink: () => void;
  };

function Result({
  className,
  result,
  onClickLink,
  titleField,
  urlField,
  thumbnailField,
  ...rest
}: ResultViewProps) {
  const fields = formatResult(result);
  const title = getEscapedField(result[titleField]);
  const url = getUrlSanitizer(URL, location.href)(getRaw(result[urlField]));
  const thumbnail = getUrlSanitizer(
    URL,
    location.href
  )(getRaw(result[thumbnailField]));

  return (
    <li className={appendClassName("sui-result", className)} {...rest}>
      <div className="sui-result__header">
        {title && !url && (
          <span
            className="sui-result__title"
            dangerouslySetInnerHTML={{ __html: title }}
          />
        )}
        {title && url && (
          <a
            className="sui-result__title sui-result__title-link"
            dangerouslySetInnerHTML={{ __html: title }}
            href={url}
            onClick={onClickLink}
            target="_blank"
            rel="noopener noreferrer"
          />
        )}
      </div>

      <div className="sui-result__body">
        {thumbnail && (
          <div className="sui-result__image">
            <img src={thumbnail} alt="" />
          </div>
        )}
        <ul className="sui-result__details">
          {Object.entries(fields).map(
            ([fieldName, fieldValue]: [string, string]) => (
              <li key={fieldName}>
                <span className="sui-result__key">{fieldName}</span>{" "}
                <span
                  className="sui-result__value"
                  dangerouslySetInnerHTML={{ __html: fieldValue }}
                />
              </li>
            )
          )}
        </ul>
      </div>
    </li>
  );
}

export default Result;
