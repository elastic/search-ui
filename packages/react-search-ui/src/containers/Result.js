import PropTypes from "prop-types";
import { Component } from "react";
import { Result } from "@elastic/react-search-ui-views";

import { withSearch } from "..";
import { Result as ResultType } from "../types";

export class ResultContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    clickThroughTags: PropTypes.arrayOf(PropTypes.string),
    titleField: PropTypes.string,
    urlField: PropTypes.string,
    thumbnailField: PropTypes.string,
    view: PropTypes.func,
    result: ResultType.isRequired,
    shouldTrackClickThrough: PropTypes.bool,
    // Actions
    trackClickThrough: PropTypes.func
  };

  static defaultProps = {
    clickThroughTags: [],
    shouldTrackClickThrough: true
  };

  handleClickLink = id => {
    const {
      clickThroughTags,
      shouldTrackClickThrough,
      trackClickThrough
    } = this.props;

    if (shouldTrackClickThrough) {
      trackClickThrough(id, clickThroughTags);
    }
  };

  render() {
    const {
      className,
      result,
      titleField,
      urlField,
      thumbnailField,
      view,
      // eslint-disable-next-line no-unused-vars
      trackClickThrough,
      // eslint-disable-next-line no-unused-vars
      shouldTrackClickThrough,
      // eslint-disable-next-line no-unused-vars
      clickThroughTags,
      ...rest
    } = this.props;
    const View = view || Result;

    return View({
      className,
      result: result,
      key: `result-${result.id.raw}`,
      onClickLink: () => this.handleClickLink(result.id.raw),
      titleField,
      urlField,
      thumbnailField,
      ...rest
    });
  }
}

export default withSearch(({ trackClickThrough }) => ({ trackClickThrough }))(
  ResultContainer
);
