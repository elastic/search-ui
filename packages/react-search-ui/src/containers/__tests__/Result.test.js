import React from "react";
import { shallow } from "enzyme";
import { ResultContainer } from "../Result";

const params = {
  result: {
    id: {
      raw: "id",
      snippet: "<em>id</em>"
    },
    title: {
      raw: "title",
      snippet: "<em>title</em>"
    },
    url: {
      raw: "url",
      snippet: "<em>url</em>"
    }
  },
  trackClickThrough: jest.fn(),
  titleField: "title",
  urlField: "url"
};

beforeEach(() => {
  params.trackClickThrough = jest.fn();
});

describe("link clicks", () => {
  it("will call back when a document link is clicked in the view", () => {
    let viewProps;

    shallow(
      <ResultContainer {...params} view={props => (viewProps = props)} />
    );

    const { onClickLink } = viewProps;
    onClickLink();

    const [id] = params.trackClickThrough.mock.calls[0];
    expect(id).toEqual("id");
  });

  it("will not call back when shouldTrackClickThrough is false", () => {
    let viewProps;

    shallow(
      <ResultContainer
        {...params}
        shouldTrackClickThrough={false}
        view={props => (viewProps = props)}
      />
    );

    const { onClickLink } = viewProps;
    onClickLink();

    expect(params.trackClickThrough.mock.calls.length).toEqual(0);
  });

  it("will pass through tags", () => {
    let viewProps;

    shallow(
      <ResultContainer
        {...params}
        clickThroughTags={["whatever"]}
        view={props => (viewProps = props)}
      />
    );

    const { onClickLink } = viewProps;
    onClickLink();

    const [id, tags] = params.trackClickThrough.mock.calls[0];
    expect(id).toEqual("id");
    expect(tags).toEqual(["whatever"]);
  });
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <ResultContainer
      {...params}
      className={className}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <ResultContainer
      {...params}
      data-foo={data}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});

it("supports a render prop", () => {
  // eslint-disable-next-line react/prop-types
  const render = ({ children }) => {
    return <div>{children}</div>;
  };
  const wrapper = shallow(<ResultContainer {...params} view={render} />);
  expect(wrapper).toMatchSnapshot();
});
