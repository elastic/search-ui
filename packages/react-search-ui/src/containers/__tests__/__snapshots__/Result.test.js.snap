// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Result
  fields={
    Object {
      "id": "<em>id</em>",
      "title": "<em>title</em>",
      "url": "<em>url</em>",
    }
  }
  key="result-id"
  onClickLink={[Function]}
  title="<em>title</em>"
  url="url"
/>
`;

exports[`supports a render prop 1`] = `<div />`;
